kythe/go/extractors/cmd/rust/rust_project_to_kzip.go

```diff
@@ -0,0 +1,552 @@
+		queue = queue[1:] // Dequeue
+
+		// Iterate over the direct dependencies of the current crate
+		for _, dep := range crateDeps[currentCrateId] {
+			// If the dependency hasn't been visited, add it to the queue and mark it as visited
+			if _, ok := visited[dep]; !ok {
+				queue = append(queue, dep)
+				visited[dep] = true
+			}
+		}
+	}
+	transitiveDeps := make([]crateId, 0)
+	for k := range visited {
+		if k != crate.CrateId {
+			transitiveDeps = append(transitiveDeps, k)
+		}
+	}
+	sort.Slice(transitiveDeps, func(i, j int) bool {
+		return transitiveDeps[i] < transitiveDeps[j]
+	})
+
+	return transitiveDeps
+}
+
+// getTransitiveDeps returns a map of all crate IDs to transitive crate dependencies
+func getTransitiveDeps(crates []crate) map[crateId][]crateId {
+	crateDeps := make(map[crateId][]crateId)
+	for _, crate := range crates {
+		for _, dep := range crate.Deps {
+			crateDeps[crate.CrateId] = append(crateDeps[crate.CrateId], dep.CrateId)
+		}
+	}
+
+	// get the transitive dependencies of each crate
+	transitiveDeps := make(map[crateId][]crateId)
+	for _, crate := range crates {
+		transitiveDepsForCrate := transitiveDepsForCrate(crate, crateDeps)
+		transitiveDeps[crate.CrateId] = transitiveDepsForCrate
+	}
+	return transitiveDeps
+}
+
+// getSourceDirs returns the set of included and excluded directories for all crates
+// Additionally, we add the directory of the root module to the set of included directories.
+// It returns a data structure containing the source directories and dependent directories, paired by include and exclude.
+// Note: the returned source directories are NOT transitive
+func getSourceDirs(crates []crate) map[crateId]source {
+	perCrateSourceDirs := make(map[crateId]source)
+	for _, crate := range crates {
+		currentSourceDirs := source{
+			IncludeDirs: []string{},
+			ExcludeDirs: []string{},
+		}
+		currentSourceDirs.IncludeDirs = append(currentSourceDirs.IncludeDirs, crate.Source.IncludeDirs...)
+		currentSourceDirs.ExcludeDirs = append(currentSourceDirs.ExcludeDirs, crate.Source.ExcludeDirs...)
+
+		// Add the dir of the root module for the crate, which is implicitly included
+		currentSourceDirs.IncludeDirs = append(currentSourceDirs.IncludeDirs, filepath.Dir(crate.RootModule))
+
+		perCrateSourceDirs[crate.CrateId] = currentSourceDirs
+	}
+
+	return perCrateSourceDirs
+}
+
+// isSubdirectory checks if the path `sub` is a subdirectory of `parent`.
+func isSubdirectory(parent, sub string) bool {
+	return !strings.HasPrefix(sub, parent)
+}
+
+// whether we should include this file in the Compilation Unit based on the inclusions/exclusions in `sourceDirs`
+func shouldIncludeFile(info os.FileInfo, path string, sourceDirs source, projectRoot string) (bool, error) {
+	if info.IsDir() {
+		// check that this file isn't in one of the dirs excluded by this crate
+		for _, excludeDir := range sourceDirs.ExcludeDirs {
+			if isSubdirectory(filepath.Join(projectRoot, excludeDir), filepath.Dir(path)) {
```

Comment:
there's no need to check subdirectory-ness, an equality check on the cleaned path is enough
(you'll never descend into the subdirectory in the first place, due to SkipDir)